src/storage/src/source/kafka.rs

```diff
@@ -671,20 +726,35 @@ impl SourceRender for KafkaSourceConnection {
                                 );
                             }
                             Ok(message) => {
-                                let (message, ts) =
-                                    construct_source_message(&message, &reader.metadata_columns);
-                                if let Some((msg, time, diff)) = reader.handle_message(message, ts)
-                                {
-                                    let pid = time.interval().singleton().unwrap().unwrap_exact();
-                                    let part_cap = &reader.partition_capabilities[pid].data;
-                                    let msg = msg.map_err(|e| {
-                                        DataflowError::SourceError(Box::new(SourceError {
-                                            error: SourceErrorDetails::Other(format!("{}", e)),
-                                        }))
-                                    });
-                                    data_output
-                                        .give_fueled(part_cap, ((0, msg), time, diff))
-                                        .await;
+                                let output_messages = outputs
+                                    .iter()
+                                    .map(|output| {
+                                        let (message, ts) = construct_source_message(
+                                            &message,
+                                            &output.metadata_columns,
+                                        );
+                                        (output.output_index, message, ts)
+                                    })
+                                    .collect::<Vec<_>>();
+                                for (output_index, message, ts) in output_messages {
+                                    if let Some((msg, time, diff)) =
```

Comment:
Unless there is something I don't see there is no reason to collect a vector here:
```suggestion
                                for output in outputs.iter() {
                                    let (message, ts) = construct_source_message(
                                        &message,
                                        &output.metadata_columns,
                                    );
                                    if let Some((msg, time, diff)) =
```

Reply:
Ah yeah this is because of a borrow-checker issue -- I should have left a comment. The `reader.handle_message` below this line requires a `&mut` on `reader` which we can't obtain while we still hold a reference to `message` which is a borrow from `reader`. I would either have to clone the `message` from the reader or do this vec allocation, so I chose the latter